Add documentations of Modern Warfare: Cubed (#28)
CleanroomMC/cleanroom-website: docs/groovy-script/mods/mwc/crafting_station.md

---

> title: "Crafting Station/Ammo Press"

Reviewer: why is this called two things? it is only modifying one set of recipes, it should be named accordingly. if those recipes can happen in multiple places, thats a job for the description, not the title

Author: It's 2 different block using same mechanism. BULLET type of recipe can only be accessed by Ammo Press, which is separate block.

Reviewer: alright, so name it something that isnt either of those options and indicate that "BULLET is in Ammo Press and everything else is in Crafting Station"

Reviewer 2: BULLET, MAGAZINE, and GRENADE are all in the ammo press 

GUN, ATTACHMENT_NORMAL, and ATTACHMENT_MODIFICATION are in the workbench

titleTemplate: "Modern Warfare Cubed | CleanroomMC"
description: "Converts multiple items into another item, with a way to recover some of original resources when dismantled."
source_code_link: "https://github.com/Cubed-Development/Modern-Warfare-Cubed/blob/next/src/main/java/com/paneedah/mwc/groovyscript/script/CraftingStation.java"
---

# Crafting Station/Ammo Press (Modern Warfare Cubed)

## Description

Converts multiple items into another item, with a way to recover some of original resources when dismantled.

## Identifier

Refer to this via any of the following:

```groovy:no-line-numbers {1}

> mods.mwc.crafting_station/* Used as page default */ // [!code focus]

Reviewer: might want to add more aliases - probably want `Alias.generateOfClassAnd(CraftingStation.class, "AmmoPress")`?

mods.mwc.craftingstation
mods.mwc.craftingStation
mods.mwc.CraftingStation
```

## Adding Recipes

### Recipe Builder

Just like other recipe types, the Crafting Station/Ammo Press also uses a recipe builder.

Don't know what a builder is? Check [the builder info page](../../getting_started/builder.md) out.

:::::::::: details mods.mwc.crafting_station.recipeBuilder() {open id="abstract"}

> - `CraftingGroup`. Category of crafting recipe. Requires not null. (Default `GUN`).

Reviewer: does not document `input` or `output`

```groovy:no-line-numbers
mods.mwc.crafting_station.recipeBuilder()
    .input(item('minecraft:clay'))
    .output(item('minecraft:diamond'))
    .register()

mods.mwc.crafting_station.recipeBuilder()
    .setYield(0.5)
    .input(item('minecraft:gold_ingot'))
    .output(item('minecraft:clay') * 2)
    .register()
```

:::::::::

::::::::::

## Removing Recipes

- Remove ALL recipes of Bullet Category:

    ```groovy:no-line-numbers
    mods.mwc.crafting_station.removeAllBullet()
    ```

- Remove ALL recipes of Gun Category:

    ```groovy:no-line-numbers
    mods.mwc.crafting_station.removeAllGun()
    ```

- Remove ALL recipes of Magazine Category:

    ```groovy:no-line-numbers
    mods.mwc.crafting_station.removeAllMagazine()
    ```

- Remove ALL recipes of Modification Attachment Category:

    ```groovy:no-line-numbers
    mods.mwc.crafting_station.removeAllModificationAttachment()
    ```

- Remove ALL recipes of Normal Attachment Category:

    ```groovy:no-line-numbers
    mods.mwc.crafting_station.removeAllNormalAttachment()
    ```

- Removes recipe that outputs matching item of given `ingredient` from bullet category:

    ```groovy:no-line-numbers

>     mods.mwc.crafting_station.removeBullet(IIngredient)

Reviewer: the standard format of method names is `removeBy[x]` - this should be `removeByBullet`. same applies to several other entries

---
title: "MWC Crafting"
titleTemplate: "Modern Warfare Cubed | CleanroomMC"
description: "Converts multiple items into another item, with a way to recover some of original resources when dismantled. recipes with group ``BULLET, MAGAZINE, GRENADE`` can be only crafted via Ammo Press, while recipes with ``GUN, ATTACHMENT_NORMAL, ATTACHMENT_MODIFICATION`` categories can only be crafted with Crafting Station."
source_code_link: "https://github.com/Cubed-Development/Modern-Warfare-Cubed/blob/next/src/main/java/com/paneedah/mwc/groovyscript/script/CraftingStation.java"
---

# MWC Crafting (Modern Warfare Cubed)

## Description

> Converts multiple items into another item, with a way to recover some of original resources when dismantled. recipes with group ``BULLET, MAGAZINE, GRENADE`` can be only crafted via Ammo Press, while recipes with ``GUN, ATTACHMENT_NORMAL, ATTACHMENT_MODIFICATION`` categories can only be crafted with Crafting Station.

Reviewer: why two backticks? only need one. might want to make it something like this:
> Any recipes with the group `BULLET`, `MAGAZINE`, or `GRENADE` can only be crafted via the Ammo Press
also, capitalization needs to be fixed. `dismantled. recipes` is incorrect.

## Identifier

Refer to this via any of the following:

```groovy:no-line-numbers {1}
mods.mwc.crafting_station/* Used as page default */ // [!code focus]
mods.mwc.craftingstation
mods.mwc.craftingStation
mods.mwc.CraftingStation
mods.mwc.ammo_press
mods.mwc.ammopress
mods.mwc.ammoPress
mods.mwc.AmmoPress
```

## Adding Recipes

### Recipe Builder

Just like other recipe types, the MWC Crafting also uses a recipe builder.

Don't know what a builder is? Check [the builder info page](../../getting_started/builder.md) out.

:::::::::: details mods.mwc.crafting_station.recipeBuilder() {open id="abstract"}

> - `IngredientList<IIngredient>`. Sets the item inputs of the recipe. Requires less than or equal to 27.

Reviewer: should also be `gte = 0` or `gte = 1` (depending).

    ```groovy:no-line-numbers
    input(IIngredient...)
    input(IIngredient, double)

>     input(double, IIngredient...)

Reviewer: be consistent about the order - have `double` be first, or dont allow varargs. might want to ditch the `input` list entirely and store `CraftingEntry` directly (with a method that is `x(double d, IIngredient i) { return x(new CraftingEntry(i.toMcIngredient(), i.getAmount(), d)); }`).

    mods.mwc.crafting_station.removeByGrenade(IIngredient)
    ```

- groovyscript.wiki.mwc.crafting_station.removeByGun:

    ```groovy:no-line-numbers
    mods.mwc.crafting_station.removeByGun(IIngredient)
    ```

- groovyscript.wiki.mwc.crafting_station.removeByMagazine:

    ```groovy:no-line-numbers
    mods.mwc.crafting_station.removeByMagazine(IIngredient)
    ```

- groovyscript.wiki.mwc.crafting_station.removeByModificationAttachment:

    ```groovy:no-line-numbers
    mods.mwc.crafting_station.removeByModificationAttachment(IIngredient)
    ```

- groovyscript.wiki.mwc.crafting_station.removeByNormalAttachment:

    ```groovy:no-line-numbers
    mods.mwc.crafting_station.removeByNormalAttachment(IIngredient)
    ```

- Removes Recipe that outputs given `ingredient` from given category. valid categories = "GUN", "ATTACHMENT_NORMAL", "ATTACHMENT_MODIFICATION", "BULLET", "MAGAZINE", "GRENADE":

    ```groovy:no-line-numbers
    mods.mwc.crafting_station.removeInGroup(IIngredient, CraftingGroup)
    ```

- Removes recipe that outputs given `ingredient` from all categories:

    ```groovy:no-line-numbers
    mods.mwc.crafting_station.remove(IIngredient)
    ```

> - groovyscript.wiki.mwc.crafting_station.removeAllInGroup:

Reviewer: no lang key (for a bunch of these)

Reviewer: no lang key still

    ```groovy:no-line-numbers
    mods.mwc.crafting_station.removeAllInGroup(CraftingGroup)
    ```

- Removes all registered recipes:

    ```groovy:no-line-numbers
    mods.mwc.crafting_station.removeAll()
    ```

:::::::::: details Example {open id="example"}
```groovy:no-line-numbers
mods.mwc.crafting_station.removeAllBullet()
mods.mwc.crafting_station.removeAllGrenade()
mods.mwc.crafting_station.removeAllGun()
mods.mwc.crafting_station.removeAllMagazine()
mods.mwc.crafting_station.removeAllModificationAttachment()
mods.mwc.crafting_station.removeAllNormalAttachment()
mods.mwc.crafting_station.removeByBullet(ore('oreDiamond'))
mods.mwc.crafting_station.removeByGrenade(ore('oreDiamond'))
mods.mwc.crafting_station.removeByGun(ore('oreDiamond'))
mods.mwc.crafting_station.removeByMagazine(ore('oreDiamond'))
mods.mwc.crafting_station.removeByModificationAttachment(ore('oreDiamond'))
mods.mwc.crafting_station.removeByNormalAttachment(ore('oreDiamond'))

> mods.mwc.crafting_station.remove(ore('oreDiamond'))

Reviewer: `remove`? remove by what? could be input, could be output. could do both in the same thing. make the method name descriptive.

---
title: "MWC Crafting"
titleTemplate: "Modern Warfare Cubed | CleanroomMC"
description: "Converts multiple items into another item, with a way to recover some of original resources when dismantled. recipes with group `BULLET, MAGAZINE, GRENADE` can be only crafted via Ammo Press, while recipes with `GUN, ATTACHMENT_NORMAL, ATTACHMENT_MODIFICATION` categories can only be crafted with Crafting Station."
source_code_link: "https://github.com/Cubed-Development/Modern-Warfare-Cubed/blob/next/src/main/java/com/paneedah/mwc/groovyscript/script/CraftingStation.java"
---

# MWC Crafting (Modern Warfare Cubed)

## Description

> Converts multiple items into another item, with a way to recover some of original resources when dismantled. recipes with group `BULLET, MAGAZINE, GRENADE` can be only crafted via Ammo Press, while recipes with `GUN, ATTACHMENT_NORMAL, ATTACHMENT_MODIFICATION` categories can only be crafted with Crafting Station.

Reviewer: recipe

## Identifier

Refer to this via any of the following:

```groovy:no-line-numbers {1}
mods.mwc.crafting_station/* Used as page default */ // [!code focus]
mods.mwc.craftingstation
mods.mwc.craftingStation
mods.mwc.CraftingStation
mods.mwc.ammo_press
mods.mwc.ammopress
mods.mwc.ammoPress
mods.mwc.AmmoPress
```

## Adding Recipes

### Recipe Builder

Just like other recipe types, the MWC Crafting also uses a recipe builder.

Don't know what a builder is? Check [the builder info page](../../getting_started/builder.md) out.

:::::::::: details mods.mwc.crafting_station.recipeBuilder() {open id="abstract"}
- `ItemStackList`. Sets the item outputs of the recipe. Requires exactly 1.

    ```groovy:no-line-numbers
    output(ItemStack)
    output(ItemStack...)
    output(Collection<ItemStack>)
    ```

- `CraftingGroup`. Category of crafting recipe. Requires not null. (Default `GUN`).

    ```groovy:no-line-numbers

>     setGroup(String)

Reviewer: want to set the priority on this down to a bit under 3000 to get this ordered in a more logical way.

    setGroupBullet()
    setGroupGrenade()
    setGroupMagazine()
    setGroupNormalAttachment()
    setGroupModificationAttachment()
    setGroup(CraftingGroup)
    ```

- `List<CraftingEntry>`. Input entries of the recipe, `double` is refund yield of the ingredient. Requires greater than 0.

    ```groovy:no-line-numbers
    input(IIngredient)
    input(IIngredient...)
    input(double, IIngredient)
    input(double, IIngredient...)
    input(Collection<IIngredient>)
    ```

- First validates the builder, returning `null` and outputting errors to the log file if the validation failed, then registers the builder and returns the registered object. (returns `null` or `com.paneedah.mwc.groovyscript.recipes.GSCraftingRecipe`).

    ```groovy:no-line-numbers
    register()
    ```

::::::::: details Example {open id="example"}
```groovy:no-line-numbers
mods.mwc.crafting_station.recipeBuilder()
    .input(1, item('minecraft:clay'))
    .output(item('minecraft:diamond'))
    .register()

mods.mwc.crafting_station.recipeBuilder()
    .input(0.5, item('minecraft:gold_ingot'))
    .output(item('minecraft:clay') * 2)
    .register()
```

:::::::::

::::::::::

## Removing Recipes

> - Remove all recipes of Bullet Category:

Reviewer: these would be better if it was rewritten to be like this
```suggestion
- Remove all recipes with the category of Bullet:
```

    ```groovy:no-line-numbers
    mods.mwc.crafting_station.removeAllBullet()
    ```

- Remove all recipes of Grenade Category:

    ```groovy:no-line-numbers
    mods.mwc.crafting_station.removeAllGrenade()
    ```

- Remove all recipes of Gun Category:

    ```groovy:no-line-numbers
    mods.mwc.crafting_station.removeAllGun()
    ```

- Remove all recipes of Magazine Category:

    ```groovy:no-line-numbers
    mods.mwc.crafting_station.removeAllMagazine()
    ```

- Remove all recipes of Modification Attachment Category:

    ```groovy:no-line-numbers
    mods.mwc.crafting_station.removeAllModificationAttachment()
    ```

- Remove all recipes of Normal Attachment Category:

    ```groovy:no-line-numbers
    mods.mwc.crafting_station.removeAllNormalAttachment()
    ```

> - Removes recipe that outputs matching item of given `ingredient` from bullet category:

Reviewer: these would be better if it was rewritten to be like this
```suggestion
- Removes all recipes that match the given output `ingredient` and are in the category Bullet:
```

    ```groovy:no-line-numbers
    mods.mwc.crafting_station.removeBulletByOutput(IIngredient)
    ```

- Removes recipe that outputs matching item of given `ingredient` from grenade category:

    ```groovy:no-line-numbers
    mods.mwc.crafting_station.removeGrenadeByOutput(IIngredient)
    ```

- Removes recipe that outputs matching item of given `ingredient` from gun category:

    ```groovy:no-line-numbers
    mods.mwc.crafting_station.removeGunByOutput(IIngredient)
    ```

> - Removes Recipe that outputs given `ingredient` from given category. valid categories = "GUN", "ATTACHMENT_NORMAL", "ATTACHMENT_MODIFICATION", "BULLET", "MAGAZINE", "GRENADE":

Reviewer: Recipe, valid.

    ```groovy:no-line-numbers

>     mods.mwc.crafting_station.removeInGroupByOutput(IIngredient, CraftingGroup)

Reviewer: probably want a `removeInGroupByOutput(IIngredient, String)` method to go along with this.

    ```

- Removes recipe that outputs matching item of given `ingredient` from magazine category:

    ```groovy:no-line-numbers
    mods.mwc.crafting_station.removeMagazineByOutput(IIngredient)
    ```

- Removes recipe that outputs matching item of given `ingredient` from modification attachment category:

    ```groovy:no-line-numbers
    mods.mwc.crafting_station.removeModificationAttachmentByOutput(IIngredient)
    ```

- Removes recipe that outputs matching item of given `ingredient` from normal attachment category:

    ```groovy:no-line-numbers
    mods.mwc.crafting_station.removeNormalAttachmentByOutput(IIngredient)
    ```

- Removes all recipes that match the given output:

    ```groovy:no-line-numbers
    mods.mwc.crafting_station.removeByOutput(IIngredient)
    ```

- groovyscript.wiki.mwc.crafting_station.removeAllInGroup:

    ```groovy:no-line-numbers

>     mods.mwc.crafting_station.removeAllInGroup(CraftingGroup)

Reviewer: would fit the standard pattern if it was renamed to be `removeByGroup`. 
additionally, probably want a `removeByGroup(String)` (`removeAllInGroup(String)`) method to go along with this.
